coffeescript-loader/loader.js

```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
   return defaultResolve(specifier, context, defaultResolve);
 }
 
-export function getFormat(url, context, defaultGetFormat) {
+export async function load(url, context, defaultLoad) {
   // Now that we patched resolve to let CoffeeScript URLs through, we need to
   // tell Node.js what format such URLs should be interpreted as. Because
   // CoffeeScript transpiles into JavaScript, it should be one of the two
   // JavaScript formats: 'commonjs' or 'module'.
   if (extensionsRegex.test(url)) {
-    // CoffeeScript files can be either CommonJS or ES modules, so we want any
-    // CoffeeScript file to be treated by Node.js the same as a .js file would
-    // be at the same location (based on the "type" field in the nearest
-    // parent package.json file). To determine how Node.js would interpret an
-    // arbitrary .js file, append .js to our CoffeeScript URL and see what
-    // format Node.js returns for such a URL. If 'commonjs' is returned, a
-    // handler will need to be registered with require.extensions to process
-    // that file via the CommonJS loader.
-    const { format } = defaultGetFormat(`${url}.js`);
-    return {format};
-  }
+    // CoffeeScript files can be either CommonJS or ES modules, but since load
+    // handles both format and source based on the same url, it cannot be used
+    // for non-js files. Instead, a quick search up the filesystem for a
+    // package.json with a `type` field settles the format issue.
+    const format = await getPackageType(dirname(fileURLToPath(url)));
 
-  // Let Node.js handle all other URLs.
-  return defaultGetFormat(url, context, defaultGetFormat);
-}
+    // source is ignored (never checked) for cjs, so safe to omit
+    if (format === 'commonjs') return { format };
 
-export function transformSource(source, context, defaultTransformSource) {
-  const { url, format } = context;
+    const { source: rawSource } = await defaultLoad(url, { format });
```

Comment:
Yes, `rawSource` is the actually (string of) CoffeeScript code here. At some point when I was debugging, I logged this, so I saw it 🙂 What's surprising? (That's what we were hoping to facilitate with `getPackageType()`)

Reply:
I was surprised that this worked because I remembered there being a check in Node’s default hook that it would always throw on unknown extensions. Maybe that’s in Node’s default `resolve` hook, not `load`? Or we took it out? I can’t remember.

Reply:
Yes, that's resolve—resolve is the bouncer at this club 🙂

Reply:
Okay, this sounds good. One more question: why are we passing `format` at all? I would think that `defaultLoad` shouldn’t need it, since its signature is the same `source, context, noop`; I’d think that just `defaultLoad(url)` should work here?

Reply:
`format` here is necessary because this loader has knowledge that Node does not (it knows to ignore the file extension and look for a package.json): omitting `format` here causes `defaultLoad()` to invoke `defaultGetFormat()` under the hood, which will fail, and thus abort before invoking `defaultGetSource()` (`defaultGetSource()` is all we actually want here, but it is not directly accessible).

Reply:
I thought we were collapsing the Node default hooks too, so that `defaultLoad` is just one function rather than a wrapper over the two legacy `defaultGetFormat` and `defaultGetSource` functions.

Reply:
We could do; I don't recall that specifically being mentioned.

When I was implementing it, keeping them separate and wrapping with defaultLoad seemed cleaner (single responsibility, separation of concerns, etc).